Fix unstable TestHelixAdminCli.testDeactivateCluster

LOG: (368)
>2020-08-21T06:26:35.4891005Z [ERROR] testDeactivateCluster(org.apache.helix.tools.TestHelixAdminCli)  Time elapsed: 2.182 s  <<< FAILURE!
2020-08-21T06:26:35.4899581Z org.apache.helix.HelixException: There are still LEADER in the cluster, shut them down first.
2020-08-21T06:26:35.4907516Z 	at org.apache.helix.tools.TestHelixAdminCli.testDeactivateCluster(TestHelixAdminCli.java:604)
2020-08-21T06:26:35.4909023Z 

>2020-08-21T04:51:16.5818512Z START testDeactivateCluster at Fri Aug 21 04:51:16 UTC 2020
2020-08-21T04:51:16.5823028Z START TestHelixAdminCli at Fri Aug 21 04:51:16 UTC 2020
2020-08-21T04:51:18.6654306Z 2360301 [HelixController-pipeline-default-TestHelixAdminCli-(f85f62bd_DEFAULT)] ERROR org.apache.helix.controller.GenericHelixController  - Exception while executing DEFAULT pipeline: TestHelixAdminCli for cluster [org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689), org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1046), org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1039), org.apache.helix.manager.zk.ZkBaseDataAccessor.getChildNames(ZkBaseDataAccessor.java:669), org.apache.helix.manager.zk.ZkBaseDataAccessor.getChildren(ZkBaseDataAccessor.java:625), org.apache.helix.manager.zk.ZkBaseDataAccessor.getChildren(ZkBaseDataAccessor.java:601), org.apache.helix.manager.zk.ZKHelixDataAccessor.getChildValues(ZKHelixDataAccessor.java:415), org.apache.helix.manager.zk.ZKHelixDataAccessor.getChildValuesMap(ZKHelixDataAccessor.java:476), org.apache.helix.common.caches.PropertyCache.doSimpleCacheRefresh(PropertyCache.java:168), org.apache.helix.common.caches.PropertyCache.refresh(PropertyCache.java:160), org.apache.helix.controller.dataproviders.BaseControllerDataProvider.doRefresh(BaseControllerDataProvider.java:333), org.apache.helix.controller.dataproviders.ResourceControllerDataProvider.refresh(ResourceControllerDataProvider.java:143), org.apache.helix.controller.stages.ReadClusterDataStage.process(ReadClusterDataStage.java:63), org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68), org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777), org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128), org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)]. Will not continue to next pipeline
2020-08-21T04:51:18.6656197Z 2360302 [HelixController-pipeline-task-TestHelixAdminCli-(252a0738_TASK)] ERROR org.apache.helix.controller.GenericHelixController  - Cluster manager: controller_9001 is not leader for TestHelixAdminCli. Pipeline will not be invoked
2020-08-21T04:51:24.8112618Z END testDeactivateCluster at Fri Aug 21 04:51:24 UTC 2020, took: 8235ms.

> 2020-09-18T00:28:03.5214953Z [ERROR] testDeactivateCluster(org.apache.helix.tools.TestHelixAdminCli)  Time elapsed: 2.244 s  <<< FAILURE!
2020-09-18T00:28:03.5216715Z org.apache.helix.HelixException: There are still LEADER in the cluster, shut them down first.
2020-09-18T00:28:03.5218782Z 	at org.apache.helix.tools.TestHelixAdminCli.testDeactivateCluster(TestHelixAdminCli.java:604)
2020-09-18T00:28:03.5220373Z 
2020-09-18T00:28:03.8965375Z [ERROR] Failures: 
2020-09-18T00:28:03.8967694Z [ERROR]   TestHelixAdminCli.testDeactivateCluster:604 » Helix There are still LEADER in ...
2020-09-18T00:28:03.8971521Z [ERROR] Tests run: 1196, Failures: 1, Errors: 0, Skipped: 2

>2020-09-17T23:34:33.0813638Z START TestHelixAdminCli testDeactivateCluster at Thu Sep 17 23:34:33 UTC 2020
2020-09-17T23:34:33.0816039Z START TestHelixAdminCli at Thu Sep 17 23:34:33 UTC 2020
2020-09-17T23:34:35.1388128Z 1254851 [HelixController-pipeline-default-TestHelixAdminCli-(041eaa31_DEFAULT)] ERROR org.apache.helix.controller.rebalancer.waged.WagedRebalancer  - Failed to calculate the new assignments.
2020-09-17T23:34:35.1399531Z org.apache.helix.HelixRebalanceException: Failed to get the current best possible assignment because of unexpected error. Failure Type: INVALID_REBALANCER_STATUS
2020-09-17T23:34:35.1403542Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:664)
2020-09-17T23:34:35.1407124Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.partialRebalance(WagedRebalancer.java:516)
2020-09-17T23:34:35.1413905Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleAssignment(WagedRebalancer.java:357)
2020-09-17T23:34:35.1419783Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleStates(WagedRebalancer.java:327)
2020-09-17T23:34:35.1424876Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeNewIdealStates(WagedRebalancer.java:262)
2020-09-17T23:34:35.1430085Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.computeResourceBestPossibleStateWithWagedRebalancer(BestPossibleStateCalcStage.java:270)
2020-09-17T23:34:35.1435441Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.compute(BestPossibleStateCalcStage.java:126)
2020-09-17T23:34:35.1438356Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.process(BestPossibleStateCalcStage.java:85)
2020-09-17T23:34:35.1440589Z 	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68)
2020-09-17T23:34:35.1442578Z 	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777)
2020-09-17T23:34:35.1447466Z 	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128)
2020-09-17T23:34:35.1449655Z 	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)
2020-09-17T23:34:35.1452127Z Caused by: org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException: java.lang.InterruptedException
2020-09-17T23:34:35.1455189Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689)
2020-09-17T23:34:35.1457227Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1826)
2020-09-17T23:34:35.1502103Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1818)
2020-09-17T23:34:35.1504733Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.get(ZkBaseDataAccessor.java:473)
2020-09-17T23:34:35.1506811Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:260)
2020-09-17T23:34:35.1509602Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:238)
2020-09-17T23:34:35.1545793Z 	at org.apache.helix.controller.rebalancer.waged.AssignmentMetadataStore.getBestPossibleAssignment(AssignmentMetadataStore.java:84)
2020-09-17T23:34:35.1549885Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:661)
2020-09-17T23:34:35.1551727Z 	... 11 more
2020-09-17T23:34:35.1552161Z Caused by: java.lang.InterruptedException
2020-09-17T23:34:35.1552733Z 	at java.lang.Object.wait(Native Method)
2020-09-17T23:34:35.1553855Z 	at java.lang.Object.wait(Object.java:502)
2020-09-17T23:34:35.1554751Z 	at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1411)
2020-09-17T23:34:35.1555901Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1219)
2020-09-17T23:34:35.1557059Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1250)
2020-09-17T23:34:35.1558424Z 	at org.apache.helix.zookeeper.zkclient.ZkConnection.readData(ZkConnection.java:144)
2020-09-17T23:34:35.1559885Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1830)
2020-09-17T23:34:35.1561067Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1826)
2020-09-17T23:34:35.1562716Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1673)
2020-09-17T23:34:35.1564014Z 	... 18 more
2020-09-17T23:34:35.1668721Z 1254851 [HelixController-pipeline-default-TestHelixAdminCli-(041eaa31_DEFAULT)] ERROR org.apache.helix.controller.stages.BestPossibleStateCalcStage  - Event 041eaa31_DEFAULT : Failed to calculate the new Ideal States using the rebalancer WagedRebalancer due to INVALID_REBALANCER_STATUS
2020-09-17T23:34:35.1678234Z org.apache.helix.HelixRebalanceException: Failed to get the current best possible assignment because of unexpected error. Failure Type: INVALID_REBALANCER_STATUS
2020-09-17T23:34:35.1723029Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:664)
2020-09-17T23:34:35.1735483Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.partialRebalance(WagedRebalancer.java:516)
2020-09-17T23:34:35.1739715Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleAssignment(WagedRebalancer.java:357)
2020-09-17T23:34:35.1749144Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleStates(WagedRebalancer.java:327)
2020-09-17T23:34:35.1752667Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeNewIdealStates(WagedRebalancer.java:262)
2020-09-17T23:34:35.1761236Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.computeResourceBestPossibleStateWithWagedRebalancer(BestPossibleStateCalcStage.java:270)
2020-09-17T23:34:35.1792270Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.compute(BestPossibleStateCalcStage.java:126)
2020-09-17T23:34:35.1795582Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.process(BestPossibleStateCalcStage.java:85)
2020-09-17T23:34:35.1797815Z 	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68)
2020-09-17T23:34:35.1808646Z 	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777)
2020-09-17T23:34:35.1811669Z 	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128)
2020-09-17T23:34:35.1814779Z 	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)
2020-09-17T23:34:35.1817938Z Caused by: org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException: java.lang.InterruptedException
2020-09-17T23:34:35.1826228Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689)
2020-09-17T23:34:35.1836738Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1826)
2020-09-17T23:34:35.1842390Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1818)
2020-09-17T23:34:35.1844666Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.get(ZkBaseDataAccessor.java:473)
2020-09-17T23:34:35.1847979Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:260)
2020-09-17T23:34:35.1850610Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:238)
2020-09-17T23:34:35.1854304Z 	at org.apache.helix.controller.rebalancer.waged.AssignmentMetadataStore.getBestPossibleAssignment(AssignmentMetadataStore.java:84)
2020-09-17T23:34:35.1862221Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:661)
2020-09-17T23:34:35.1864232Z 	... 11 more
2020-09-17T23:34:35.1864690Z Caused by: java.lang.InterruptedException
2020-09-17T23:34:35.1865304Z 	at java.lang.Object.wait(Native Method)
2020-09-17T23:34:35.1865836Z 	at java.lang.Object.wait(Object.java:502)
2020-09-17T23:34:35.1866769Z 	at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1411)
2020-09-17T23:34:35.1867960Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1219)
2020-09-17T23:34:35.1869006Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1250)
2020-09-17T23:34:35.1870458Z 	at org.apache.helix.zookeeper.zkclient.ZkConnection.readData(ZkConnection.java:144)
2020-09-17T23:34:35.1872015Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1830)
2020-09-17T23:34:35.1873445Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1826)
2020-09-17T23:34:35.1875222Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1673)
2020-09-17T23:34:35.1878217Z 	... 18 more
2020-09-17T23:34:35.1946437Z 1254852 [HelixController-pipeline-default-TestHelixAdminCli-(041eaa31_DEFAULT)] ERROR org.apache.helix.controller.GenericHelixController  - Exception while executing DEFAULT pipeline: TestHelixAdminCli for cluster [org.apache.helix.zookeeper.zkclient.ZkClient.acquireEventLock(ZkClient.java:1629), org.apache.helix.zookeeper.zkclient.ZkClient.waitForKeeperState(ZkClient.java:1607), org.apache.helix.zookeeper.zkclient.ZkClient.waitUntilConnected(ZkClient.java:1598), org.apache.helix.manager.zk.ZKHelixManager.checkConnected(ZKHelixManager.java:396), org.apache.helix.manager.zk.ZKHelixManager.getSessionId(ZKHelixManager.java:943), org.apache.helix.controller.stages.MessageGenerationPhase.createStateTransitionMessage(MessageGenerationPhase.java:412), org.apache.helix.controller.stages.MessageGenerationPhase.generateMessage(MessageGenerationPhase.java:235), org.apache.helix.controller.stages.MessageGenerationPhase.processEvent(MessageGenerationPhase.java:102), org.apache.helix.controller.stages.resource.ResourceMessageGenerationPhase.process(ResourceMessageGenerationPhase.java:35), org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68), org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777), org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128), org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)]. Will not continue to next pipeline
2020-09-17T23:34:35.2024765Z 1254852 [HelixController-pipeline-task-TestHelixAdminCli-(562e8de3_TASK)] ERROR org.apache.helix.controller.GenericHelixController  - Exception while executing TASK pipeline: TestHelixAdminCli for cluster [org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689), org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1046), org.apache.helix.zookeeper.zkclient.ZkClient.getChildren(ZkClient.java:1039), org.apache.helix.manager.zk.ZkBaseDataAccessor.getChildNames(ZkBaseDataAccessor.java:669), org.apache.helix.manager.zk.ZKHelixDataAccessor.getChildNames(ZKHelixDataAccessor.java:394), org.apache.helix.common.caches.InstanceMessagesCache.refresh(InstanceMessagesCache.java:113), org.apache.helix.controller.dataproviders.BaseControllerDataProvider.doRefresh(BaseControllerDataProvider.java:341), org.apache.helix.controller.dataproviders.WorkflowControllerDataProvider.refresh(WorkflowControllerDataProvider.java:88), org.apache.helix.controller.stages.ReadClusterDataStage.process(ReadClusterDataStage.java:63), org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68), org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777), org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128), org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)]. Will not continue to next pipeline
2020-09-17T23:34:35.4207164Z 1255134 [HelixController-pipeline-default-TestHelixAdminCli-(f12c5deb_DEFAULT)] ERROR org.apache.helix.controller.GenericHelixController  - Cluster manager: controller_9000 is not leader for TestHelixAdminCli. Pipeline will not be invoked
2020-09-17T23:34:41.3800177Z END TestHelixAdminCli testDeactivateCluster at Thu Sep 17 23:34:41 UTC 2020, took: 8298ms.

2020-09-24T02:01:42.8529030Z [ERROR] testDeactivateCluster(org.apache.helix.tools.TestHelixAdminCli)  Time elapsed: 7.763 s  <<< FAILURE!
2020-09-24T02:01:42.8537470Z org.apache.helix.zookeeper.exception.ZkClientException: Failed to delete /TestHelixAdminCli
2020-09-24T02:01:42.8546009Z 	at org.apache.helix.tools.TestHelixAdminCli.testDeactivateCluster(TestHelixAdminCli.java:615)
2020-09-24T02:01:42.8560841Z Caused by: org.apache.helix.zookeeper.zkclient.exception.ZkException: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /TestHelixAdminCli
2020-09-24T02:01:42.8570363Z 	at org.apache.helix.tools.TestHelixAdminCli.testDeactivateCluster(TestHelixAdminCli.java:615)
2020-09-24T02:01:42.8580752Z Caused by: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /TestHelixAdminCli
2020-09-24T02:01:42.8589069Z 	at org.apache.helix.tools.TestHelixAdminCli.testDeactivateCluster(TestHelixAdminCli.java:615)
2020-09-24T02:01:42.8591548Z 
2020-09-24T02:01:43.2439422Z [ERROR] Failures: 
2020-09-24T02:01:43.2442720Z [ERROR]   TestHelixAdminCli.testDeactivateCluster:615 » ZkClient Failed to delete /TestH...

>2020-09-23T05:08:49.1861208Z START TestHelixAdminCli at Wed Sep 23 05:08:49 UTC 2020
2020-09-23T05:08:49.3335448Z 1197589 [ZkClient-EventThread-25630-localhost:2183] ERROR org.apache.helix.zookeeper.zkclient.callback.ZkAsyncCallbacks  - Interrupted waiting for success
2020-09-23T05:08:49.3337814Z java.lang.InterruptedException
2020-09-23T05:08:49.3338699Z 	at java.lang.Object.wait(Native Method)
2020-09-23T05:08:49.3339546Z 	at java.lang.Object.wait(Object.java:502)
2020-09-23T05:08:49.3341383Z 	at org.apache.helix.zookeeper.zkclient.callback.ZkAsyncCallbacks$DefaultCallback.waitForSuccess(ZkAsyncCallbacks.java:220)
2020-09-23T05:08:49.3343677Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.issueSync(ZkClient.java:1305)
2020-09-23T05:08:49.3345356Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.access$300(ZkClient.java:83)
2020-09-23T05:08:49.3346813Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$4.run(ZkClient.java:1331)
2020-09-23T05:08:49.3348377Z 	at org.apache.helix.zookeeper.zkclient.ZkEventThread.run(ZkEventThread.java:99)
2020-09-23T05:08:53.3085745Z 1201556 [HelixController-pipeline-default-TestHelixAdminCli-(51359ebb_DEFAULT)] ERROR org.apache.helix.zookeeper.zkclient.callback.ZkAsyncCallbacks  - Interrupted waiting for success
2020-09-23T05:08:53.3088355Z java.lang.InterruptedException
2020-09-23T05:08:53.3088980Z 	at java.lang.Object.wait(Native Method)
2020-09-23T05:08:53.3089527Z 	at java.lang.Object.wait(Object.java:502)
2020-09-23T05:08:53.3091244Z 	at org.apache.helix.zookeeper.zkclient.callback.ZkAsyncCallbacks$DefaultCallback.waitForSuccess(ZkAsyncCallbacks.java:220)
2020-09-23T05:08:53.3093597Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.create(ZkBaseDataAccessor.java:761)
2020-09-23T05:08:53.3095719Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.createChildren(ZkBaseDataAccessor.java:831)
2020-09-23T05:08:53.3098099Z 	at org.apache.helix.manager.zk.ZKHelixDataAccessor.createChildren(ZKHelixDataAccessor.java:515)
2020-09-23T05:08:53.3100729Z 	at org.apache.helix.controller.stages.MessageDispatchStage.sendMessages(MessageDispatchStage.java:171)
2020-09-23T05:08:53.3103834Z 	at org.apache.helix.controller.stages.MessageDispatchStage.processEvent(MessageDispatchStage.java:81)
2020-09-23T05:08:53.3107326Z 	at org.apache.helix.controller.stages.resource.ResourceMessageDispatchStage.process(ResourceMessageDispatchStage.java:33)
2020-09-23T05:08:53.3110066Z 	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68)
2020-09-23T05:08:53.3112170Z 	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777)
2020-09-23T05:08:53.3114571Z 	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128)
2020-09-23T05:08:53.3116852Z 	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)
2020-09-23T05:08:54.8285053Z testDeactivateCluster deavtive cluster TestHelixAdminCli
2020-09-23T05:08:54.9471743Z 1203203 [HelixController-pipeline-default-TestHelixAdminCli-(4ebbd400_DEFAULT)] ERROR org.apache.helix.controller.GenericHelixController  - Cluster manager: controller_9000 is not leader for TestHelixAdminCli. Pipeline will not be invoked
2020-09-23T05:08:54.9490029Z 1203203 [HelixController-pipeline-task-TestHelixAdminCli-(4ebbd400_TASK)] ERROR org.apache.helix.controller.GenericHelixController  - Cluster manager: controller_9000 is not leader for TestHelixAdminCli. Pipeline will not be invoked
2020-09-23T05:08:54.9860669Z validated leader path not existing TestHelixAdminCli
2020-09-23T05:08:55.1822844Z 1203434 [HelixController-pipeline-default-TestHelixAdminCli_grand-(012e378a_DEFAULT)] ERROR org.apache.helix.controller.rebalancer.waged.WagedRebalancer  - Failed to calculate the new assignments.
2020-09-23T05:08:55.1827373Z org.apache.helix.HelixRebalanceException: Failed to get the current best possible assignment because of unexpected error. Failure Type: INVALID_REBALANCER_STATUS
2020-09-23T05:08:55.1831747Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:664)
2020-09-23T05:08:55.1836423Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.partialRebalance(WagedRebalancer.java:516)
2020-09-23T05:08:55.1840246Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleAssignment(WagedRebalancer.java:357)
2020-09-23T05:08:55.1844462Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleStates(WagedRebalancer.java:327)
2020-09-23T05:08:55.1848257Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeNewIdealStates(WagedRebalancer.java:262)
2020-09-23T05:08:55.1853838Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.computeResourceBestPossibleStateWithWagedRebalancer(BestPossibleStateCalcStage.java:270)
2020-09-23T05:08:55.1858904Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.compute(BestPossibleStateCalcStage.java:126)
2020-09-23T05:08:55.1862144Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.process(BestPossibleStateCalcStage.java:85)
2020-09-23T05:08:55.1864973Z 	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68)
2020-09-23T05:08:55.1867222Z 	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777)
2020-09-23T05:08:55.1869747Z 	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128)
2020-09-23T05:08:55.1872174Z 	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)
2020-09-23T05:08:55.1876644Z Caused by: org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException: java.lang.InterruptedException
2020-09-23T05:08:55.1879279Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689)
2020-09-23T05:08:55.1924731Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1826)
2020-09-23T05:08:55.1926458Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1818)
2020-09-23T05:08:55.1928993Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.get(ZkBaseDataAccessor.java:473)
2020-09-23T05:08:55.1931382Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:260)
2020-09-23T05:08:55.1934256Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:238)
2020-09-23T05:08:55.1938125Z 	at org.apache.helix.controller.rebalancer.waged.AssignmentMetadataStore.getBestPossibleAssignment(AssignmentMetadataStore.java:84)
2020-09-23T05:08:55.2008862Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:661)
2020-09-23T05:08:55.2010934Z 	... 11 more
2020-09-23T05:08:55.2011416Z Caused by: java.lang.InterruptedException
2020-09-23T05:08:55.2012782Z 	at java.lang.Object.wait(Native Method)
2020-09-23T05:08:55.2013697Z 	at java.lang.Object.wait(Object.java:502)
2020-09-23T05:08:55.2014758Z 	at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1411)
2020-09-23T05:08:55.2016030Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1219)
2020-09-23T05:08:55.2017136Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1250)
2020-09-23T05:08:55.2018677Z 	at org.apache.helix.zookeeper.zkclient.ZkConnection.readData(ZkConnection.java:144)
2020-09-23T05:08:55.2020326Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1830)
2020-09-23T05:08:55.2021654Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1826)
2020-09-23T05:08:55.2023473Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1673)
2020-09-23T05:08:55.2024714Z 	... 18 more
2020-09-23T05:08:55.2033258Z 1203434 [HelixController-pipeline-default-TestHelixAdminCli_grand-(012e378a_DEFAULT)] ERROR org.apache.helix.controller.stages.BestPossibleStateCalcStage  - Event 012e378a_DEFAULT : Failed to calculate the new Ideal States using the rebalancer WagedRebalancer due to INVALID_REBALANCER_STATUS
2020-09-23T05:08:55.2045999Z org.apache.helix.HelixRebalanceException: Failed to get the current best possible assignment because of unexpected error. Failure Type: INVALID_REBALANCER_STATUS
2020-09-23T05:08:55.2055735Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:664)
2020-09-23T05:08:55.2065610Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.partialRebalance(WagedRebalancer.java:516)
2020-09-23T05:08:55.2076790Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleAssignment(WagedRebalancer.java:357)
2020-09-23T05:08:55.2096796Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeBestPossibleStates(WagedRebalancer.java:327)
2020-09-23T05:08:55.2107688Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.computeNewIdealStates(WagedRebalancer.java:262)
2020-09-23T05:08:55.2122412Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.computeResourceBestPossibleStateWithWagedRebalancer(BestPossibleStateCalcStage.java:270)
2020-09-23T05:08:55.2134962Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.compute(BestPossibleStateCalcStage.java:126)
2020-09-23T05:08:55.2145058Z 	at org.apache.helix.controller.stages.BestPossibleStateCalcStage.process(BestPossibleStateCalcStage.java:85)
2020-09-23T05:08:55.2155306Z 	at org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68)
2020-09-23T05:08:55.2165092Z 	at org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777)
2020-09-23T05:08:55.2174779Z 	at org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128)
2020-09-23T05:08:55.2184115Z 	at org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)
2020-09-23T05:08:55.2193459Z Caused by: org.apache.helix.zookeeper.zkclient.exception.ZkInterruptedException: java.lang.InterruptedException
2020-09-23T05:08:55.2222030Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1689)
2020-09-23T05:08:55.2228830Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1826)
2020-09-23T05:08:55.2235286Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.readData(ZkClient.java:1818)
2020-09-23T05:08:55.2245941Z 	at org.apache.helix.manager.zk.ZkBaseDataAccessor.get(ZkBaseDataAccessor.java:473)
2020-09-23T05:08:56.0029773Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:260)
2020-09-23T05:08:56.0033520Z 	at org.apache.helix.manager.zk.ZkBucketDataAccessor.compressedBucketRead(ZkBucketDataAccessor.java:238)
2020-09-23T05:08:56.0037707Z 	at org.apache.helix.controller.rebalancer.waged.AssignmentMetadataStore.getBestPossibleAssignment(AssignmentMetadataStore.java:84)
2020-09-23T05:08:56.0042324Z 	at org.apache.helix.controller.rebalancer.waged.WagedRebalancer.getBestPossibleAssignment(WagedRebalancer.java:661)
2020-09-23T05:08:56.0050241Z 	... 11 more
2020-09-23T05:08:56.0050749Z Caused by: java.lang.InterruptedException
2020-09-23T05:08:56.0051390Z 	at java.lang.Object.wait(Native Method)
2020-09-23T05:08:56.0051951Z 	at java.lang.Object.wait(Object.java:502)
2020-09-23T05:08:56.0052949Z 	at org.apache.zookeeper.ClientCnxn.submitRequest(ClientCnxn.java:1411)
2020-09-23T05:08:56.0055097Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1219)
2020-09-23T05:08:56.0056257Z 	at org.apache.zookeeper.ZooKeeper.getData(ZooKeeper.java:1250)
2020-09-23T05:08:56.0057802Z 	at org.apache.helix.zookeeper.zkclient.ZkConnection.readData(ZkConnection.java:144)
2020-09-23T05:08:56.0059455Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1830)
2020-09-23T05:08:56.0060788Z 	at org.apache.helix.zookeeper.zkclient.ZkClient$11.call(ZkClient.java:1826)
2020-09-23T05:08:56.0062608Z 	at org.apache.helix.zookeeper.zkclient.ZkClient.retryUntilConnected(ZkClient.java:1673)
2020-09-23T05:08:56.0063839Z 	... 18 more
2020-09-23T05:08:56.0079934Z 1203444 [HelixController-pipeline-default-TestHelixAdminCli_grand-(012e378a_DEFAULT)] ERROR org.apache.helix.controller.GenericHelixController  - Exception while executing DEFAULT pipeline: TestHelixAdminCli_grand for cluster [org.apache.helix.zookeeper.zkclient.ZkClient.acquireEventLock(ZkClient.java:1629), org.apache.helix.zookeeper.zkclient.ZkClient.waitForKeeperState(ZkClient.java:1607), org.apache.helix.zookeeper.zkclient.ZkClient.waitUntilConnected(ZkClient.java:1598), org.apache.helix.manager.zk.ZKHelixManager.checkConnected(ZKHelixManager.java:396), org.apache.helix.manager.zk.ZKHelixManager.getHelixDataAccessor(ZKHelixManager.java:658), org.apache.helix.controller.stages.MessageDispatchStage.processEvent(MessageDispatchStage.java:64), org.apache.helix.controller.stages.resource.ResourceMessageDispatchStage.process(ResourceMessageDispatchStage.java:33), org.apache.helix.controller.pipeline.Pipeline.handle(Pipeline.java:68), org.apache.helix.controller.GenericHelixController.handleEvent(GenericHelixController.java:777), org.apache.helix.controller.GenericHelixController.access$500(GenericHelixController.java:128), org.apache.helix.controller.GenericHelixController$ClusterEventProcessor.run(GenericHelixController.java:1407)]. Will not continue to next pipeline
2020-09-23T05:08:56.0096249Z 1203488 [HelixController-pipeline-task-TestHelixAdminCli_grand-(b2c13aed_TASK)] ERROR org.apache.helix.controller.GenericHelixController  - Cluster manager: controller_9001 is not leader for TestHelixAdminCli_grand. Pipeline will not be invoked
2020-09-23T05:08:56.0098411Z END TestHelixAdminCli testDeactivateCluster at Wed Sep 23 05:08:55 UTC 2020, took: 6073ms.

--- a/helix-core/src/test/java/org/apache/helix/tools/TestHelixAdminCli.java
+++ b/helix-core/src/test/java/org/apache/helix/tools/TestHelixAdminCli.java
@@ -29,6 +29,7 @@ import java.util.Set;
 import org.apache.helix.BaseDataAccessor;
 import org.apache.helix.HelixDataAccessor;
 import org.apache.helix.TestHelper;
+import org.apache.helix.tools.ClusterVerifiers.BestPossibleExternalViewVerifier;
 import org.apache.helix.zookeeper.datamodel.ZNRecord;
 import org.apache.helix.common.ZkTestBase;
 import org.apache.helix.integration.manager.ClusterDistributedController;
@@ -568,7 +569,14 @@ public class TestHelixAdminCli extends ZkTestBase {
     String command =
         "-zkSvr " + ZK_ADDR + " -activateCluster " + clusterName + " " + grandClusterName + " true";
     ClusterSetup.processCommandLineArgs(command.split("\\s+"));
-    Thread.sleep(500);
+    // wait till grand_cluster converge
+    BestPossibleExternalViewVerifier verifier = new BestPossibleExternalViewVerifier.Builder(grandClusterName)
+        .setZkClient(_gZkClient)
+        .setWaitTillVerify(TestHelper.DEFAULT_REBALANCE_PROCESSING_WAIT_TIME)
+        .build();
+
+    boolean result = verifier.verifyByPolling();
+    Assert.assertTrue(result, "grand cluster not converging.");
 
     // deactivate cluster
     command = "-zkSvr " + ZK_ADDR + " -activateCluster " + clusterName + " " + grandClusterName
@@ -602,6 +610,12 @@ public class TestHelixAdminCli extends ZkTestBase {
 
     command = "-zkSvr localhost:2183 -dropCluster " + clusterName;
     ClusterSetup.processCommandLineArgs(command.split("\\s"));
+
+    boolean leaderNotExists = TestHelper.verify(() -> {
+      boolean isLeaderExists = _gZkClient.exists(path);
+      return isLeaderExists == false;
+    }, 10000L);
+    Assert.assertTrue(leaderNotExists, " mysterious leader out!");
 
     for (ClusterDistributedController controller : controllers) {
       controller.syncStop();

--- a/helix-core/src/test/java/org/apache/helix/tools/TestHelixAdminCli.java
+++ b/helix-core/src/test/java/org/apache/helix/tools/TestHelixAdminCli.java
@@ -614,7 +614,7 @@ public class TestHelixAdminCli extends ZkTestBase {
     boolean leaderNotExists = TestHelper.verify(() -> {
       boolean isLeaderExists = _gZkClient.exists(path);
       return isLeaderExists == false;
-    }, 10000L);
+    }, TestHelper.WAIT_DURATION);
     Assert.assertTrue(leaderNotExists, " mysterious leader out!");
 
     for (ClusterDistributedController controller : controllers) {

